crates/ursa-pod/src/codec.rs

```diff
@@ -1,2 +1,543 @@
+use arrayref::array_ref;
+use bytes::{BufMut, Bytes, BytesMut};
 use tokio_util::codec::{Decoder, Encoder};
 
+pub type EpochNonce = u64;
+pub type Secp256k1AffinePoint = [u8; 33];
+
+pub type Blake3CID = [u8; 32];
+
+pub type Secp256k1PublicKey = Secp256k1AffinePoint;
+pub type SchnorrSignature = [u8; 64];
+
+pub type BLSPublicKey = [u8; 48];
+pub type BLSSignature = [u8; 96];
+
+// Compression algorithm bitmap
+pub const NONE: u8 = 1 << 0;
+pub const SNAPPY: u8 = 1 << 1;
+pub const GZIP: u8 = 1 << 2;
+pub const LZ4: u8 = 1 << 3;
+
+pub const NETWORK: [u8; 4] = *b"URSA";
+pub const MAX_FRAME_SIZE: usize = 1024;
+pub const MAX_LANES: u8 = 24;
+
+// Frame Tags
+pub const REQ: u8 = 0b00000000;
+pub const RES: u8 = 0b10000000;
+pub const HANDSHAKE_REQ_TAG: u8 = 0x00;
+pub const HANDSHAKE_RES_TAG: u8 = 0x80;
+pub const CONTENT_REQ_TAG: u8 = REQ | 0x01;
+pub const CONTENT_RES_TAG: u8 = RES | 0x01;
+pub const RANGE_REQ_TAG: u8 = REQ | 0x02;
+pub const UPDATE_EPOCH_TAG: u8 = RES | 0x02;
+pub const DECRYPTION_KEY_REQ_TAG: u8 = REQ | 0x03;
+pub const DECRYPTION_KEY_RES_TAG: u8 = RES | 0x03;
+pub const END_OF_REQUEST_TAG: u8 = 0xFE;
+pub const TERMINATE_TAG: u8 = 0xFF;
+
+#[derive(Default)]
+pub struct UrsaCodec {}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Blake3Tree {}
+
+impl Blake3Tree {
+    fn to_bytes(&self) -> Bytes {
+        Bytes::from_static(b"blake3tree")
+    }
+
+    fn from_bytes(_bytes: &mut BytesMut) -> Self {
+        Self {}
+    }
+}
+
+#[repr(u8)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Reason {
+    InsufficientBalance = 0x00,
+    Unknown = 0xFF,
+}
+
+impl Reason {
+    fn from_u8(byte: u8) -> Option<Self> {
+        match byte {
+            0x00 => Some(Self::InsufficientBalance),
+            0xFF => Some(Self::Unknown),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct LastLaneData {}
+
+/// Frame variants for different requests and responses
+#[repr(u8)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum UrsaFrame {
```

Comment:
don't comment on the exact tag number here anymore since we have the constant, instead just having `TAG . ...` is sufficient.